config.org
#+STARTUP: overview

* Configurações gerais
** Iniciando o emacs
#+BEGIN_SRC emacs-lisp
;; Disable startup screen
(setq inhibit-splash-screen t)
#+END_SRC
** Ajustes úteis
#+BEGIN_SRC emacs-lisp
;; https://github.com/danielmai/.emacs.d/blob/master/config.org

;; quebra de linha automática
(global-visual-line-mode t)

;; These functions are useful. Activate them.
(put 'downcase-region 'disabled nil)
(put 'upcase-region 'disabled nil)
(put 'narrow-to-region 'disabled nil)
(put 'dired-find-alternate-file 'disabled nil)

;; Answering just 'y' or 'n' will do
(defalias 'yes-or-no-p 'y-or-n-p)

;; Keep all backup and auto-save files in one directory
(setq backup-directory-alist '(("." . "~/.emacs.d/backups")))
(setq auto-save-file-name-transforms '((".*" "~/.emacs.d/auto-save-list/" t)))

;; UTF-8 please
(setq locale-coding-system 'utf-8) ; pretty
(set-terminal-coding-system 'utf-8) ; pretty
(set-keyboard-coding-system 'utf-8) ; pretty
(set-selection-coding-system 'utf-8) ; please
(prefer-coding-system 'utf-8) ; with sugar on top

;; delete the region when typing, just like as we expect nowadays.
(delete-selection-mode t)

;; avisa se parênteses estão em número igual de abertura e fechamento
(show-paren-mode t)

(column-number-mode t)

;; Don't beep at me
(setq visible-bell t)

#+END_SRC
** Tema
#+BEGIN_SRC emacs-lisp
;; Tema escuro
(load-theme 'wombat t)
#+END_SRC

** Informações pessoais
#+BEGIN_SRC emacs-lisp
(setq user-full-name "Rafael Rodrigues de Moraes"
      user-mail-address "hafer.moraes@gmail.com")

;; chaves de acesso aos bancos de dados
(org-babel-load-file (expand-file-name "~/.emacs.d/sql-connector.org"))
#+END_SRC

** Configurações específicas 
Ao invés de poluir o init.el com detalhes de configurações específicas, usa-se o arquivo =custom.el= para guardá-las.
#+BEGIN_SRC emacs-lisp
(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(load custom-file)

(setq auto-mode-alist (cons '("\\.tex$" . latex-mode) auto-mode-alist)) ; https://tex.stackexchange.com/questions/275794/emacs-is-not-recognizing-files-as-latex-and-going-into-latex-mode
(setq LaTeX-indent-level 4) 
(setq LaTeX-item-indent 2)
#+END_SRC


* Correção ortográfica e gramática
É necessário instalar os pacotes =aspell= e =aspell-pt-br= no sistema.
 
#+BEGIN_SRC emacs-lisp
(setq ispell-dictionary "brasileiro")

;; easy spell check (fonte: https://www.emacswiki.org/emacs/FlySpell)
(global-set-key (kbd "<f8>") 'ispell-word)
(global-set-key (kbd "<f9>") 'ispell-region)

;; LanguageTool
;(setq langtool-language-tool-jar "/home/claurafa/.emacs.d/languagetool/LanguageTool-4.4/languagetool-commandline.jar")
;(require 'langtool)
;(setq langtool-default-language "pt-BR")
;(setq langtool-mother-tongue "pt")

;(defun langtool-autoshow-detail-popup (overlays)
;  (when (require 'popup nil t)
;    ;; Do not interrupt current popup
;    (unless (or popup-instances
;                ;; suppress popup after type `C-g` .
;                (memq last-command '(keyboard-quit)))
;      (let ((msg (langtool-details-error-message overlays)))
;        (popup-tip msg)))))
;(setq langtool-autoshow-message-function
;      'langtool-autoshow-detail-popup)
#+END_SRC


* Autocomplete
Recurso de completar comandos com base nas letras iniciais.
#+BEGIN_SRC emacs-lisp
;; Autocomplete
(use-package auto-complete
  :ensure t)
;; (use-package auto-complete-config
;;   :ensure t)
(ac-config-default)
(global-auto-complete-mode t)
#+END_SRC


* ESS e Outline-mode (folding para R)
Facilitar o trabalho com a linguagem R
#+BEGIN_SRC emacs-lisp
;; ESS
(use-package ess
  :ensure t)
(setq ess-use-auto-complete t)
(setq ess-ask-for-ess-directory nil)
#+END_SRC

Folding de comentários de scripts do R
#+BEGIN_SRC emacs-lisp
;;; Outline-magic
;(add-hook 'outline-mode-hook 
;          (lambda () 
;            (require 'outline-cycle)))
(use-package outline-magic
  :ensure t)
(add-hook 'outline-minor-mode-hook 
          (lambda () 
            (require 'outline-magic)
            (define-key outline-minor-mode-map  (kbd "<C-tab>") 'outline-cycle)))
(add-hook 'ess-mode-hook
      '(lambda ()
         (outline-minor-mode)
         (setq outline-regexp "#\\{3,5\\} ")
         (defun outline-level ()
           (cond ((looking-at "^### "      ) 1)
		 ((looking-at "^#### "     ) 2)
		 ((looking-at "^##### "    ) 3)
		 ((looking-at "^###### "   ) 4)
		 ;((looking-at "^[a-zA-Z0-9_\.]+ ?<- ?function(.*{") 3)
		 (t 1000)))
         ))
#+END_SRC


* RMarkdown
Configurações para tornar o emacs um editor similar ao RStudio e para trabalhar bem com arquivos .Rmd
#+BEGIN_SRC emacs-lisp
;; R-Markdown
(use-package markdown-mode
  :ensure t)
(use-package poly-R
  :ensure t)
;;(require 'poly-R)
;;(require 'poly-markdown)
;;(add-to-list 'auto-mode-alist '("\\.Rmd\'" . poly-markdown+r-mode))
;;(setq load-path (append ’(“/home/claurafa/.emacs.d/polymode/” “/home/claurafa/.emacs.d/polymode/modes”) load-path))
#+END_SRC


* Org
** Configuracoes gerais
#+BEGIN_SRC emacs-lisp
(use-package org
  :ensure org-plus-contrib)
;(require 'org)


(setq org-image-actual-width 550)

(setq org-highlight-latex-and-related '(latex script entities))

(setq org-hide-emphasis-markers t)

(setq org-export-default-language "pt")

(setq org-support-shift-select (quote always))

#+END_SRC

** interleave-mode
Tomar notas de livros e pdfs
#+BEGIN_SRC emacs-lisp
(use-package interleave
  :ensure t)

  (setq interleave-split-direction (quote vertical)
	interleave-split-lines 20
	interleave-disable-narrowing t
	)
  
#+END_SRC
** Org-babel
Blocos de código em arquivos =.org=
#+BEGIN_SRC emacs-lisp
  ;; Org-babel
  (org-babel-do-load-languages
   'org-babel-load-languages
   '( (R . t)
      (org . t)
      (ledger . t)
      (sql . t)
      (shell . t)
      (sqlite . t)
      ))

  (defun my-org-confirm-babel-evaluate (lang body)
    "Do not confirm evaluation for these languages."
    (not (or (string= lang "R")
	     (string= lang "shell")
	     (string= lang "python")
	     (string= lang "emacs-lisp")
	     (string= lang "sqlite"))))
  (setq org-confirm-babel-evaluate 'my-org-confirm-babel-evaluate)

(setq org-confirm-babel-evaluate nil
      org-src-window-setup 'current-window
      org-src-strip-leading-and-trailing-blank-lines t
      org-src-preserve-indentation t
      org-src-fontify-natively t
      org-src-tab-acts-natively t)
#+END_SRC

** Org-capture
Para anotações rápidas e direcionadas em arquivos específicos
#+BEGIN_SRC emacs-lisp
;; Org-Capture
(setq org-default-notes-file (concat org-directory "/inbox.org"))
(define-key global-map (kbd "<f7>") 'org-capture)
;; Templates do Org-Capture
(setq org-capture-templates
      '(("t" "Todo [Inbox]" entry
	 (file+headline "~/gtd/inbox.org" "Tasks")
        "* TODO %^{Título}\n %^{Breve descritivo da tarefa}\n  %U \n\n")
	("j" "Journal" entry
	 (file+olp+datetree "~/gtd/journal.org")
	 "* %?\nEntered on %U\n  %i\n  %a")
	))
(setq org-agenda-files '("~/gtd/inbox.org"
                         "~/gtd/notes.org"))
(setq org-refile-targets '(("~/gtd/inbox.org" :maxlevel . 3)
                           ("~/gtd/notes.org" :maxlevel . 2)))
#+END_SRC
** Org-latex
Configuracoes específicas para a compilacao de documentos LaTeX.
#+BEGIN_SRC emacs-lisp
  (setq org-latex-default-packages-alist
	(quote
	 (("AUTO" "inputenc" t)
	  ("T1" "fontenc" t)
	  ("hidelinks" "hyperref" nil)
	  "\\tolerance=1000")))

  (setq org-latex-listings 'minted
	org-latex-packages-alist '(("" "minted"))
	org-latex-pdf-process
	'("pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"
	  "bibtex %b"
	  "pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"
	  "pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"))
  (setq org-latex-minted-options
	'(
	  ("frame" "lines")
	  ("linenos=true")
	  ("fontsize=\\footnotesize")
	  )
	)
#+END_SRC
*** pdflatex
#+BEGIN_SRC emacs-lisp
;; pdflatex
(require 'ox-latex)
(unless (boundp 'org-latex-classes)
  (setq org-latex-classes nil))
(add-to-list 'org-latex-classes
             '("beamer"
               "\\documentclass\[presentation\]\{beamer\}"
               ("\\section\{%s\}" . "\\section*\{%s\}")
               ("\\subsection\{%s\}" . "\\subsection*\{%s\}")
               ("\\subsubsection\{%s\}" . "\\subsubsection*\{%s\}"))
             '("article"
               "\\documentclass{article}"
               ("\\section{%s}" . "\\section*{%s}")
               ("\\subsection{%s}" . "\\subsection*{%s}")
               ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
               ("\\paragraph{%s}" . "\\paragraph*{%s}")
               ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))
(setq org-latex-to-pdf-process '("texi2dvi --pdf --clean --verbose --batch %f"))
#+END_SRC


* reveal.js (apresentações)
Apresentacoes baseadas em arquivos =html=. Necessita de configuracao adicional e um pouco de trabalho.
#+BEGIN_SRC emacs-lisp
;; reveal.js
(use-package ox-reveal
  :ensure t)
#+END_SRC


* PDFTools
Anotação de arquivos pdf e cores mais escuras para tornar a leitura mais agradável.
#+BEGIN_SRC emacs-lisp
;; pdf-tools e midnight mode
(use-package pdf-tools
  :ensure t)
(add-hook 'pdf-view-mode-hook (lambda ()
				(pdf-view-midnight-minor-mode))) ; automatically turns on midnight-mode for pdfs
(setq pdf-view-midnight-colors '("light gray" . "gray14" )) ; set the amber profile as default (see below)
(pdf-tools-install)
#+END_SRC




config.el
;; Disable startup screen
(setq inhibit-splash-screen t)

;; https://github.com/danielmai/.emacs.d/blob/master/config.org

;; quebra de linha automática
(global-visual-line-mode t)

;; These functions are useful. Activate them.
(put 'downcase-region 'disabled nil)
(put 'upcase-region 'disabled nil)
(put 'narrow-to-region 'disabled nil)
(put 'dired-find-alternate-file 'disabled nil)

;; Answering just 'y' or 'n' will do
(defalias 'yes-or-no-p 'y-or-n-p)

;; Keep all backup and auto-save files in one directory
(setq backup-directory-alist '(("." . "~/.emacs.d/backups")))
(setq auto-save-file-name-transforms '((".*" "~/.emacs.d/auto-save-list/" t)))

;; UTF-8 please
(setq locale-coding-system 'utf-8) ; pretty
(set-terminal-coding-system 'utf-8) ; pretty
(set-keyboard-coding-system 'utf-8) ; pretty
(set-selection-coding-system 'utf-8) ; please
(prefer-coding-system 'utf-8) ; with sugar on top

;; delete the region when typing, just like as we expect nowadays.
(delete-selection-mode t)

;; avisa se parênteses estão em número igual de abertura e fechamento
(show-paren-mode t)

(column-number-mode t)

;; Don't beep at me
(setq visible-bell t)

;; Tema escuro
(load-theme 'wombat t)

(setq user-full-name "Rafael Rodrigues de Moraes"
      user-mail-address "hafer.moraes@gmail.com")

(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(load custom-file)

(setq ispell-dictionary "brasileiro")

;; easy spell check (fonte: https://www.emacswiki.org/emacs/FlySpell)
(global-set-key (kbd "<f8>") 'ispell-word)
(global-set-key (kbd "<f9>") 'ispell-region)

;; LanguageTool
;(setq langtool-language-tool-jar "/home/claurafa/.emacs.d/languagetool/LanguageTool-4.4/languagetool-commandline.jar")
;(require 'langtool)
;(setq langtool-default-language "pt-BR")
;(setq langtool-mother-tongue "pt")

;(defun langtool-autoshow-detail-popup (overlays)
;  (when (require 'popup nil t)
;    ;; Do not interrupt current popup
;    (unless (or popup-instances
;                ;; suppress popup after type `C-g` .
;                (memq last-command '(keyboard-quit)))
;      (let ((msg (langtool-details-error-message overlays)))
;        (popup-tip msg)))))
;(setq langtool-autoshow-message-function
;      'langtool-autoshow-detail-popup)

;; Autocomplete
(use-package auto-complete
  :ensure t)
;; (use-package auto-complete-config
;;   :ensure t)
(ac-config-default)
(global-auto-complete-mode t)

;; ESS
(setq ess-use-auto-complete t)

;;; Outline-magic
;(add-hook 'outline-mode-hook 
;          (lambda () 
;            (require 'outline-cycle)))
(add-hook 'outline-minor-mode-hook 
          (lambda () 
            (require 'outline-magic)
            (define-key outline-minor-mode-map  (kbd "<C-tab>") 'outline-cycle)))
(add-hook 'ess-mode-hook
      '(lambda ()
         (outline-minor-mode)
         (setq outline-regexp "#\\{3,5\\} ")
         (defun outline-level ()
           (cond ((looking-at "^### "      ) 1)
		 ((looking-at "^#### "     ) 2)
		 ((looking-at "^##### "    ) 3)
		 ((looking-at "^###### "   ) 4)
		 ;((looking-at "^[a-zA-Z0-9_\.]+ ?<- ?function(.*{") 3)
		 (t 1000)))
         ))

;; R-Markdown
(use-package markdown-mode
  :ensure t)
;;(require 'poly-R)
;;(require 'poly-markdown)
;;(add-to-list 'auto-mode-alist '("\\.Rmd\'" . poly-markdown+r-mode))
;;(setq load-path (append ’(“/home/claurafa/.emacs.d/polymode/” “/home/claurafa/.emacs.d/polymode/modes”) load-path))

(use-package org
  :ensure org-plus-contrib)
;(require 'org)


(setq org-image-actual-width 550)

(setq org-highlight-latex-and-related '(latex script entities))

(setq org-hide-emphasis-markers t)

(setq org-export-default-language "pt")

(setq org-support-shift-select (quote always))

(setq interleave-split-direction (quote vertical)
      interleave-split-lines 20
      interleave-disable-narrowing t
      )

;; Org-babel
  (org-babel-do-load-languages
   'org-babel-load-languages
   '( (R . t)
      (org . t)
      (ledger . t)
      (sql . t)
      (shell . t)
      (sqlite . t)
      ))

  (defun my-org-confirm-babel-evaluate (lang body)
    "Do not confirm evaluation for these languages."
    (not (or (string= lang "R")
	     (string= lang "shell")
	     (string= lang "python")
	     (string= lang "emacs-lisp")
	     (string= lang "sqlite"))))
  (setq org-confirm-babel-evaluate 'my-org-confirm-babel-evaluate)

(setq org-confirm-babel-evaluate nil
      org-src-window-setup 'current-window
      org-src-strip-leading-and-trailing-blank-lines t
      org-src-preserve-indentation t
      org-src-fontify-natively t
      org-src-tab-acts-natively t)

;; Org-Capture
(setq org-default-notes-file (concat org-directory "/inbox.org"))
(define-key global-map (kbd "<f7>") 'org-capture)
;; Templates do Org-Capture
(setq org-capture-templates
      '(("t" "Todo [Inbox]" entry
	 (file+headline "~/org/inbox.org" "Tasks")
        "* TODO %^{Título}\n %^{Breve descritivo da tarefa}\n  %U \n\n")
	("j" "Journal" entry
	 (file+olp+datetree "~/org/journal.org")
	 "* %?\nEntered on %U\n  %i\n  %a")
	))
(setq org-agenda-files '("~/org/inbox.org"
                         "~/org/notes.org"))
(setq org-refile-targets '(("~/org/inbox.org" :maxlevel . 3)
                           ("~/org/notes.org" :maxlevel . 2)))

(setq org-latex-default-packages-alist
      (quote
       (("AUTO" "inputenc" t)
	("T1" "fontenc" t)
	("hidelinks" "hyperref" nil)
	"\\tolerance=1000")))

(setq org-latex-listings 'minted
      org-latex-packages-alist '(("" "minted"))
      org-latex-pdf-process
      '("pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"
	"bibtex %b"
	"pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"
	"pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"))
(setq org-latex-minted-options
      '(
	("frame" "lines")
	("linenos=true")
	("fontsize=\\footnotesize")
	)
      )

;; pdflatex
(require 'ox-latex)
(unless (boundp 'org-latex-classes)
  (setq org-latex-classes nil))
(add-to-list 'org-latex-classes
             '("beamer"
               "\\documentclass\[presentation\]\{beamer\}"
               ("\\section\{%s\}" . "\\section*\{%s\}")
               ("\\subsection\{%s\}" . "\\subsection*\{%s\}")
               ("\\subsubsection\{%s\}" . "\\subsubsection*\{%s\}"))
             '("article"
               "\\documentclass{article}"
               ("\\section{%s}" . "\\section*{%s}")
               ("\\subsection{%s}" . "\\subsection*{%s}")
               ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
               ("\\paragraph{%s}" . "\\paragraph*{%s}")
               ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))
(setq org-latex-to-pdf-process '("texi2dvi --pdf --clean --verbose --batch %f"))

;; reveal.js
(use-package ox-reveal
  :ensure t)

;; pdf-tools e midnight mode
(add-hook 'pdf-view-mode-hook (lambda ()
				(pdf-view-midnight-minor-mode))) ; automatically turns on midnight-mode for pdfs
(setq pdf-view-midnight-colors '("light gray" . "gray14" )) ; set the amber profile as default (see below)
(pdf-tools-install)
